### ISSUE TITLE
insert_batch returns -1 when a query error is encountered

### ISSUE THREAD

**Issue by sneakyimp** [bcit-ci/CodeIgniter#4925]
I've been trying to work with insert_batch and have been getting a return value of -1 when my parameters result in a query error. According to the docs, this should result in a return value of FALSE:
http://www.codeigniter.com/userguide3/database/query_builder.html#CI_DB_query_builder::insert_batch

Here's a proof-of-concept controller method:
```php
	public function barf() {
		$data = array(
				array(
						'title' => 'My title',
						'name' => 'My Name',
						'date' => 'My date'
				),
				array(
						'title' => 'Another title',
						'name' => 'Another Name',
						'date' => 'Another date'
				)
		);
		$retval = $this->db->insert_batch("i_do_not_exist", $data);
		var_dump($retval);
	}
```

You should also be able to reproduce this problem by trying to insert records with a primary key value that conflicts with some existing primary key (or unique key).

I haven't done exhaustive testing, but it would appear that any query error will result in a return value of -1.

**Comment by narfbg** [comment]
This would imply that `query()` returns `TRUE` or a non-boolean, non-empty value on failure ... Which driver is that?

**Comment by sneakyimp** [comment]
MySQL. My workstation is running 5.6.

**Comment by narfbg** [comment]
driver != platform

We have mysql, mysqli, pdo_mysql ...

**Comment by sneakyimp** [comment]
mysqli looks like?
```php
$db['default'] = array(
	'dsn'	=> '',
	'hostname' => 'localhost',
	'username' => '<username>',
	'password' => '<password>',
	'database' => '<database-name>',
	'dbdriver' => 'mysqli',
	'dbprefix' => '',
	'pconnect' => FALSE,
	'db_debug' => (ENVIRONMENT !== 'production'),
	'cache_on' => FALSE,
	'cachedir' => '',
	'char_set' => 'utf8',
	'dbcollat' => 'utf8_general_ci',
	'swap_pre' => '',
	'encrypt' => FALSE,
	'compress' => FALSE,
	'stricton' => FALSE,
	'failover' => array(),
);
```

**Comment by narfbg** [comment]
Can't reproduce.

In fact, unless you're running this in a 'production' environment, you'd get no return value at all - a fatal error will be triggered.

**Comment by vlakoff** [comment]
In `insert_batch()`:
```php
$affected_rows = 0;
// query loop
    $affected_rows += $this->affected_rows();
```
Then the mysql driver calls [`mysql_affected_rows()`](http://php.net/manual/en/function.mysql-affected-rows.php), which returns `-1` on failure.

**Comment by narfbg** [comment]
I saw that, but it requires that `query()` would return boolean TRUE - I'm not aware of a scenario where that would happen (even if implicitly casted) AND then `affected_rows()` would return -1. Doesn't make any sense either.

**Comment by vlakoff** [comment]
I couldn't figure it out either.

**Comment by sneakyimp** [comment]
I apologize. I *think* the problem is fixed in d9a4063f87d82836e7d7fe340c5c962e0332e2e5 when you switched this:
```php
		for ($i = 0, $total = count($this->qb_set); $i < $total; $i += $batch_size)
		{
			$this->query($this->_insert_batch($this->protect_identifiers($table, TRUE, $escape, FALSE), $this->qb_keys, array_slice($this->qb_set, $i, $batch_size)));
			$affected_rows += $this->affected_rows();
		}
```
to this:
```php
		for ($i = 0, $total = count($this->qb_set); $i < $total; $i += $batch_size)
		{
			if ($this->query($this->_insert_batch($this->protect_identifiers($table, TRUE, $escape, FALSE), $this->qb_keys, array_slice($this->qb_set, $i, $batch_size))))
			{
				$affected_rows += $this->affected_rows();
			}
		}
```

Problem *was* issue between the [mysqli driver](https://github.com/bcit-ci/CodeIgniter/blob/3.0-stable/system/database/drivers/mysqli/mysqli_driver.php#L402). The affected_rows function returns a -1 under [certain circumstances](http://php.net/manual/en/mysqli.affected-rows.php):
>An integer greater than zero indicates the number of rows affected or retrieved. Zero indicates that no records were updated for an UPDATE statement, no rows matched the WHERE clause in the query or that no query has yet been executed. -1 indicates that the query returned an error.

The old code that doesn't check the result of query() was returning that -1.

I cannot figure out when this code made it into the 3.0 branch. Not sure how I missed updating this file.  Will update my CI and confirm when I get a moment.

**Comment by vlakoff** [comment]
The fix has been applied to the 3.0 branch at the same time, but there has been no new tag in 3.0 since then.

**Comment by vlakoff** [comment]
Mmmm, what am I saying. The fix has been applied to the 3.0 branch, and the first release to have it is 3.1.0.

(see my confusion? 3.0 branch, 3.1 releases on it :p)

**Comment by narfbg** [comment]
*sigh*

> I cannot figure out when this code made it into the 3.0 branch. Not sure how I missed updating this file.

... this can only happen if you update every single file by hand, separately. Don't do this.

---

Logging that this is a duplicate of #4713.
